FEATURE: Asset usage strategies

radmiraal opened the pull request:
This change allows to register asset usage strategies that
define if an asset is still in use and how many times.

radmiraal commented:
tests of the media package still directly test the repository, this might have to be updated too

aertmann commented on TYPO3.Neos/Classes/TYPO3/Neos/Domain/Service/AssetService.php:
```diff
@@ -0,0 +1,67 @@
+
+        $relatedNodes = $this->nodeDataRepository->findNodesByRelatedEntities($relationMap);
+        if (count($relatedNodes) > 0) {
+            throw new \Exception('Asset could not be deleted, because there are still Nodes using it.');
```
needs a code

aertmann commented on TYPO3.Neos/Classes/TYPO3/Neos/Domain/Service/AssetService.php:
```diff
@@ -0,0 +1,67 @@
+    protected $nodeDataRepository;
+
+    /**
+     * @param AssetInterface $asset
```
@return void

aertmann commented on TYPO3.Media/Classes/TYPO3/Media/Domain/Service/AssetServiceInterface.php:
```diff
@@ -0,0 +1,26 @@
+
+    public function update(AssetInterface $asset);
+
+    public function add(AssetInterface $asset);
```
would order them differently.. add, update, remove

aertmann commented on TYPO3.Media/Classes/TYPO3/Media/Domain/Service/AssetServiceInterface.php:
```diff
@@ -0,0 +1,26 @@
+ */
+interface AssetServiceInterface
+{
+    public function remove(AssetInterface $asset);
```
needs annotations

aertmann commented on TYPO3.Media/Classes/TYPO3/Media/Domain/Service/AssetService.php:
```diff
@@ -42,6 +48,67 @@ class AssetService
+    /**
+     * Add an asset object.
+     *
+     * @param AssetInterface $asset
```
@return void

aertmann commented on TYPO3.Media/Classes/TYPO3/Media/Domain/Service/AssetService.php:
```diff
@@ -42,6 +48,67 @@ class AssetService
+    /**
+     * Update an asset object.
+     *
+     * @param AssetInterface $asset
```
@return void

aertmann commented on TYPO3.Media/Classes/TYPO3/Media/Domain/Service/AssetService.php:
```diff
@@ -42,6 +48,67 @@ class AssetService
+    public function add(AssetInterface $asset)
+    {
+        $this->getRepository($asset)->add($asset);
+    }
```
would order them differently.. add, update, remove

aertmann commented:
Hey @radmiraal, good change just found some easy nitpicks but functionally looks fine.

aertmann commented:
argh sorry, it's marked WIP.. that mention bot brings attention before it's ready :/

radmiraal commented:
lol ;-) No, thanks a lot for the feedback... actually the WIP was just because I need to rework the tests, not regarding the code itself :-) So will process the feedback while fixing the tests

radmiraal commented:
Ready for review, @robertlemke could use your attention too

aertmann commented on TYPO3.Neos/Classes/TYPO3/Neos/Domain/Service/LegacySiteImportService.php:
```diff
@@ -457,11 +464,7 @@ protected function importNodeProperty(\SimpleXMLElement $nodePropertyXml, NodeIn
-                    } else {
-                        $this->assetRepository->$repositoryAction($entry);
-                    }
+                    $this->assetService->$repositoryAction($entry);
```
$repositoryAction should probably be renamed

aertmann commented on TYPO3.Media/Classes/TYPO3/Media/Domain/Service/AssetService.php:
```diff
@@ -42,6 +49,65 @@ class AssetService
+     * Update an asset object.
+     *
+     * @param AssetInterface $asset
+     * @throws \TYPO3\Flow\Persistence\Exception\IllegalObjectTypeException
```
does it?

aertmann commented on TYPO3.Media/Classes/TYPO3/Media/Domain/Service/AssetService.php:
```diff
@@ -42,6 +49,65 @@ class AssetService
+     * Add an asset object.
+     *
+     * @param AssetInterface $asset
+     * @throws \TYPO3\Flow\Persistence\Exception\IllegalObjectTypeException
```
does it?

aertmann commented:
reviewed again, found a couple of new nitpicks but 👍 with those fixed

radmiraal commented:
fixed the nitpicks, thanks for reviewing!

aertmann commented on TYPO3.Neos/Classes/TYPO3/Neos/Domain/Service/LegacySiteImportService.php:
```diff
@@ -456,11 +463,10 @@ protected function importNodeProperty(\SimpleXMLElement $nodePropertyXml, NodeIn
+                        $this->assetService->update($entry);
                     } else {
-                        $this->assetRepository->$repositoryAction($entry);
+                        $this->assetService->add($entry);
```
whoops this is reversed now right?

radmiraal replied:
nope, all repository 'write' actions are done at the service. Just looks like reversal because I reintroduced the if statement and removed the ternary :)

robertlemke commented:
Hey Rens, thanks for the PR! As usual, I'll turn into challenging mode now, to make sure that we really want to introduce this ...

So, could it be that, what you really want - instead of a service - is a versatile repository? I could imagine that, instead of deprecating the repository access (which will certainly be ignored for some time) and introducing a new service and interface, we could introduce the functionality into the AssetRepository itself. If you override `add`, `remove` and so on, so that it checks first if the object should go into a specialized repository, people could still just use `AssetRepository` for all their asset storage needs.

Your turn to challenge my idea!

radmiraal commented:
Hey Robert, thanks ;-)

First of all I had the exact same thoughts while coding. I did extend the already existing service because I needed a place for a follow up change which I don't think belongs to the repository. As you can see in 
https://github.com/radmiraal/neos-development-collection/pull/1/commits/b5a596564dda8232a5013a549d42d795c0f610a4

You're completely right that people will work around the service, I thought of ways to force usage of the service but can't think of any accept running ahead and go cqrs-y again... ;)

Having the service is btw also a possible extension point to prevent certain actions from being executed, like creating assets with duplicate names... It's kind of finding a way in current architecture while it would be so cool to have cqrs / event sourcing in place already.

So I can not really challenge your idea because I do agree to your points fully, I just don't see a better solution for the usage references (this logic was in the controller before...) / blocking commands atm.

robertlemke commented:
All right. What if we keep the `AssetRepository` as the main responsible, but enhance its `add`, `update`... methods so that it uses the new `AssetService`?

That would mean, you could simply take any `$asset` and pass it to `add()` of the `AssetRepository`. But instead of adding it directly, the `AssetRepository` would work together with the `AssetService` internally. The `AssetRepository` could then reject new assets based on what the `AssetService` says.

It's a bit intransparent compared to dealing with the `AssetService` directly, but on the other hand, it is not such a big change compared to the previous API.

radmiraal commented:
ok, I can agree to that, but then we end up in the funny situation I need to merge the usage strategies in this branch because it otherwise doesn't make much sense anymore ;) Will fix

radmiraal commented:
- removed the AssetInterface
- added the check for usage to the repository remove() method
- merged the asset usage strategy change into it

@aertmann can you remove your thumb and review again please? As the change now has more functionality

daniellienert commented on TYPO3.Media/Classes/TYPO3/Media/Controller/AssetController.php:
```diff
@@ -392,8 +401,13 @@ public function addAssetToCollectionAction(Asset $asset, AssetCollection $assetC
-        $this->addFlashMessage(sprintf('Asset "%s" has been deleted.', htmlspecialchars($asset->getLabel())));
+        try {
+            $this->assetRepository->remove($asset);
+            $this->addFlashMessage(sprintf('Asset "%s" has been deleted.', htmlspecialchars($asset->getLabel())));
```
Would be great to have these messages translatable.

radmiraal replied:
Yes, will take care of that in https://github.com/neos/neos-development-collection/pull/464/ but it's just a move of a line and not really related to the PR. So if you're fine with me fixing it in the other PR I'd love to see a thumb ;)

aertmann replied:
👍 should actually be fixed in 2.0 and up

radmiraal replied:
see https://github.com/neos/neos-development-collection/pull/502

aertmann commented on TYPO3.Media/Classes/TYPO3/Media/Domain/Model/Asset.php:
```diff
@@ -413,4 +420,24 @@ public function setAssetCollections(\Doctrine\Common\Collections\Collection $ass
+    /**
+     * Returns the number of times the asset is in use.
+     *
+     * @return integer
```
should these be marked with `@api`?

radmiraal replied:
thought about that too, will add it as it makes sense

aertmann commented on TYPO3.Neos/Classes/TYPO3/Neos/Domain/Strategy/AssetUsageInNodePropertiesStrategy.php:
```diff
@@ -0,0 +1,138 @@
+/**
+ * @Flow\Scope("singleton")
+ */
+class AssetUsageInNodePropertiesStrategy extends AbstractAssetUsageStrategy implements AssetUsageStrategyInterface
```
`implements AssetUsageStrategyInterface` is redundant

aertmann commented on TYPO3.Media/Documentation/DeveloperInformation/index.rst:
```diff
@@ -0,0 +1,61 @@
+	/**
+	 * @Flow\Scope("singleton")
+	 */
+	class MyCustomAssetUsageStrategy extends AbstractAssetUsageStrategy implements AssetUsageStrategyInterface
```
`implements AssetUsageStrategyInterface` is redundant

aertmann commented on TYPO3.Media/Classes/TYPO3/Media/Domain/Repository/AssetRepository.php:
```diff
@@ -278,4 +284,13 @@ public function findAllIterator()
     }
+
+    /**
+     * @param AssetInterface $object
```
@return void

aertmann commented on TYPO3.Neos/Classes/TYPO3/Neos/Domain/Model/Dto/AssetUsageInNodeProperties.php:
```diff
@@ -0,0 +1,90 @@
+    public function isAccessible()
+    {
+        return $this->accessible;
+    }
```
would probably have left some of this info out, but since everything is fetched anyway except the document node it makes sense

aertmann commented on TYPO3.Media/Classes/TYPO3/Media/Domain/Repository/AssetRepository.php:
```diff
@@ -278,4 +284,13 @@ public function findAllIterator()
+    {
+        $this->assetService->validateRemoval($object);
+        parent::remove($object);
+    }
```
now this enforces the policy that usages cannot be removed, is there any need to be able to force it? since there's no way to remove the usages easily..

radmiraal replied:
I thought about adding a second argument to the remove() method indeed, but not sure about adding boolean arguments that way... Maybe a forceRemove($object) method...

aertmann commented:
@radmiraal: nicely done, looks really good and certainly changes the existing API much less.. left some questions/comments

radmiraal commented:
processed the comments, should be ready to merge now

aertmann commented:
looks good 👍 from my side

robertlemke commented:
Great work, thanks a lot for taking the extra mile!

I think I spotted a bug though - @radmiraal would you try to verify this one? Should work with any website, but I give the instruction right how I can reproduce it:

- use the current Neos Demo site
- go to the Media module an upload a new image
- go to some page, for example "Features"
- use the Media Browser for setting the new image as the background image and press apply
- the new image is now the background image
- click "Remove" in the inspector in order to remove the image again, then press Apply
- a 500er will occur, the message found in the JSON string is: `"Exception while property mapping for target type "TYPO3\TYPO3CR\Domain\Model\Node", at property path "": Asset could not be deleted, because it is still in use."` (code 1297759968)
- reload the UI
- click "Remove" and Apply again
- no error will occur, but the image won't be removed either

radmiraal commented:
Pushed a fix, also test what happens in the following usecase:

* User A adds an image and uses it in the site
* User A publishes to live
* User B changes the alttext of the image but does not publish
* User A removes the image using the inspector and publishes

This successfully removes the image from User A's workspace, and does not break user B's workspace, he still sees the image and can publish the change (which will of course display the image again).

robertlemke commented:
Thanks for the fix!

This works for me now, except one weird thing which is also happening in master right now (and thus is unrelated to this change): when you try to remove an asset which is still in use, you don't get a flash message anymore.

But as for this one, I think we can merge it.

robertlemke commented:
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;BpmQu96F&#x2F;2-display-asset-references"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> Display asset references</a>
